File: photo-studio/app/gemini_client.py
"""Google Gemini image-editing integration, with strict product-preservation
instructions, retry/timeout/rate-limit handling, caching and a safety gate.

If the ``google-genai`` package or a network call is unavailable, or the API
key is missing, ``GeminiClient.is_configured`` is False and callers should
use the local Pillow/OpenCV fallback instead — this module never raises on
"not configured", it just reports itself unusable.
"""
from __future__ import annotations

import hashlib
import io
import time
from dataclasses import dataclass
from pathlib import Path

from PIL import Image

from app.config import Settings
from app.image_processing.local_processor import resize_max_dimension
from app.models import ProcessingMode
from app.similarity import structural_similarity

try:
    from google import genai
    from google.genai import types as genai_types
    from google.genai import errors as genai_errors

    GENAI_IMPORT_ERROR: Exception | None = None
except Exception as exc:  # pragma: no cover - exercised only if package missing
    genai = None
    genai_types = None
    genai_errors = None
    GENAI_IMPORT_ERROR = exc


MODE_FOCUS = {
    ProcessingMode.CLEAN_PRODUCT_PHOTO: "Improve lighting, background and framing for a clean catalog look.",
    ProcessingMode.BACKGROUND_CLEANUP: "Focus only on cleaning up and neutralising the background.",
    ProcessingMode.LIGHTING_QUALITY: "Focus only on lighting, exposure, contrast, white balance and sharpness. Do not change the background or framing.",
    ProcessingMode.MARKETPLACE_COVER: "Create a clean, centred cover-image presentation suitable for a marketplace listing.",
    ProcessingMode.BATCH_CONSISTENCY: "Match a consistent, neutral studio presentation style suitable for a batch of similar product photos.",
}

BASE_INSTRUCTION = (
    "Improve the presentation of this clothing product photograph. Preserve the exact garment, "
    "colour, shape, size, texture, logos, labels and all visible defects. You may improve lighting, "
    "framing and the background only. Do not remove or conceal stains, holes, tears, damage or wear. "
    "Do not add or remove product details. Do not change the pattern, print or fabric texture. "
    "The result must remain an honest, accurate representation of the original product photograph."
)


def build_instruction(mode: ProcessingMode, background_style: str) -> str:
    focus = MODE_FOCUS.get(mode, "")
    background_note = ""
    if background_style != "preserve_original":
        background_note = f" Use a simple, neutral {background_style.replace('_', ' ')} background."
    else:
        background_note = " Keep the original background as-is."
    return f"{BASE_INSTRUCTION} {focus}{background_note}"


class GeminiRateLimitError(Exception):
    pass


class GeminiTimeoutError(Exception):
    pass


class GeminiRequestError(Exception):
    pass


@dataclass
class GeminiEditResult:
    image: Image.Image | None
    used_cache: bool
    from_gemini: bool
    ssim: float | None
    is_safe: bool
    error: str | None = None


class GeminiClient:
    def __init__(self, settings: Settings, cache_dir: Path | None = None):
        self.settings = settings
        self.cache_dir = cache_dir
        if self.cache_dir:
            self.cache_dir.mkdir(parents=True, exist_ok=True)
        self._client = None
        if self.is_configured:
            self._client = genai.Client(api_key=settings.gemini_api_key)

    @property
    def is_configured(self) -> bool:
        return bool(self.settings.gemini_api_key) and genai is not None

    def _cache_key(self, image_bytes: bytes, instruction: str) -> str:
        h = hashlib.sha256()
        h.update(image_bytes)
        h.update(instruction.encode("utf-8"))
        h.update(self.settings.gemini_image_model.encode("utf-8"))
        return h.hexdigest()

    def _cache_path(self, key: str) -> Path | None:
        if not self.cache_dir:
            return None
        return self.cache_dir / f"{key}.png"

    def _call_api_once(self, image: Image.Image, instruction: str) -> bytes:
        buf = io.BytesIO()
        image.save(buf, format="PNG")
        image_part = genai_types.Part.from_bytes(data=buf.getvalue(), mime_type="image/png")

        response = self._client.models.generate_content(
            model=self.settings.gemini_image_model,
            contents=[instruction, image_part],
            config=genai_types.GenerateContentConfig(
                http_options=genai_types.HttpOptions(timeout=self.settings.gemini_timeout_seconds * 1000),
            ),
        )

        candidates = getattr(response, "candidates", None) or []
        for candidate in candidates:
            content = getattr(candidate, "content", None)
            if not content:
                continue
            for part in getattr(content, "parts", None) or []:
                inline_data = getattr(part, "inline_data", None)
                if inline_data is not None and inline_data.data:
                    return inline_data.data
        raise GeminiRequestError("Gemini response did not contain an edited image.")

    def _call_with_retries(self, image: Image.Image, instruction: str) -> bytes:
        last_exc: Exception | None = None
        for attempt in range(1, self.settings.gemini_max_retries + 1):
            try:
                return self._call_api_once(image, instruction)
            except Exception as exc:  # noqa: BLE001 - classify below
                last_exc = exc
                message = str(exc).lower()
                is_rate_limit = "rate" in message or "429" in message or "resource_exhausted" in message
                is_timeout = "timeout" in message or "deadline" in message
                if attempt >= self.settings.gemini_max_retries:
                    break
                backoff = self.settings.gemini_retry_backoff_seconds * (2 ** (attempt - 1))
                if is_rate_limit:
                    time.sleep(backoff * 2)
                elif is_timeout:
                    time.sleep(backoff)
                else:
                    time.sleep(backoff)
        assert last_exc is not None
        raise last_exc

    def edit_image(self, image: Image.Image, mode: ProcessingMode, background_style: str) -> GeminiEditResult:
        if not self.is_configured:
            return GeminiEditResult(
                image=None, used_cache=False, from_gemini=False, ssim=None, is_safe=False,
                error="Gemini is not configured (missing API key or package)."
            )

        instruction = build_instruction(mode, background_style)
        upload_image = resize_max_dimension(image, self.settings.gemini_max_upload_dimension)

        buf = io.BytesIO()
        upload_image.save(buf, format="PNG")
        cache_key = self._cache_key(buf.getvalue(), instruction)
        cache_path = self._cache_path(cache_key)

        if cache_path and cache_path.exists():
            edited = Image.open(cache_path).convert("RGB")
            ssim = structural_similarity(upload_image, edited)
            is_safe = ssim >= self.settings.gemini_min_ssim
            return GeminiEditResult(image=edited, used_cache=True, from_gemini=True, ssim=ssim, is_safe=is_safe)

        try:
            raw_bytes = self._call_with_retries(upload_image, instruction)
            edited = Image.open(io.BytesIO(raw_bytes)).convert("RGB")
        except Exception as exc:  # noqa: BLE001
            return GeminiEditResult(
                image=None, used_cache=False, from_gemini=False, ssim=None, is_safe=False, error=str(exc)
            )

        ssim = structural_similarity(upload_image, edited)
        is_safe = ssim >= self.settings.gemini_min_ssim

        if cache_path:
            try:
                edited.save(cache_path, format="PNG")
            except OSError:
                pass  # cache write failures are non-fatal

        return GeminiEditResult(image=edited, used_cache=False, from_gemini=True, ssim=ssim, is_safe=is_safe)


def estimate_api_usage_warning(num_images: int, use_gemini: bool) -> str | None:
    if not use_gemini or num_images == 0:
        return None
    return (
        f"This batch will send up to {num_images} image edit request(s) to the Gemini API "
        "(fewer if some are served from cache). Large batches may take a while and are subject "
        "to your Google AI Studio / Gemini API usage quota and billing."
    )

File: photo-studio-php/includes/gemini_client.php
<?php
declare(strict_types=1);

/**
 * Google Gemini image-editing integration via the REST API (cURL — no SDK
 * needed, works on any shared host with the curl extension enabled).
 *
 * Strict product-preservation instructions are sent with every request,
 * and every edited image is compared against the original with a coarse
 * perceptual-similarity check before being trusted (see
 * jps_image_similarity() below). If it's not configured (no API key), or a
 * call fails, callers should use the local GD fallback instead — this
 * module never throws for "not configured", it just reports itself unusable.
 */

const JPS_MODE_FOCUS = [
    'clean_product_photo' => 'Improve lighting, background and framing for a clean catalog look.',
    'background_cleanup' => 'Focus only on cleaning up and neutralising the background.',
    'lighting_quality' => 'Focus only on lighting, exposure, contrast, white balance and sharpness. Do not change the background or framing.',
    'marketplace_cover' => 'Create a clean, centred cover-image presentation suitable for a marketplace listing.',
    'batch_consistency' => 'Match a consistent, neutral studio presentation style suitable for a batch of similar product photos.',
];

const JPS_GEMINI_BASE_INSTRUCTION =
    'Improve the presentation of this clothing product photograph. Preserve the exact garment, ' .
    'colour, shape, size, texture, logos, labels and all visible defects. You may improve lighting, ' .
    'framing and the background only. Do not remove or conceal stains, holes, tears, damage or wear. ' .
    'Do not add or remove product details. Do not change the pattern, print or fabric texture. ' .
    'The result must remain an honest, accurate representation of the original product photograph.';

function jps_build_gemini_instruction(string $mode, string $backgroundStyle): string
{
    $focus = JPS_MODE_FOCUS[$mode] ?? '';
    if ($backgroundStyle !== 'preserve_original') {
        $label = str_replace('_', ' ', $backgroundStyle);
        $backgroundNote = " Use a simple, neutral $label background.";
    } else {
        $backgroundNote = ' Keep the original background as-is.';
    }
    return JPS_GEMINI_BASE_INSTRUCTION . ' ' . $focus . $backgroundNote;
}

/**
 * Coarse perceptual similarity (0..1, 1 = identical) between two GD images.
 * Not true SSIM (no OpenCV available in plain PHP) — a 32x32 grayscale
 * average-difference comparison, good enough as a safety gate to reject
 * Gemini edits that look structurally very different from the source.
 */
function jps_image_similarity(\GdImage $a, \GdImage $b, int $gridSize = 32): float
{
    $grayA = jps_grayscale_grid($a, $gridSize);
    $grayB = jps_grayscale_grid($b, $gridSize);

    $n = count($grayA);
    $sumDiff = 0.0;
    for ($i = 0; $i < $n; $i++) {
        $sumDiff += abs($grayA[$i] - $grayB[$i]);
    }
    $avgDiff = $n > 0 ? $sumDiff / $n : 255.0;
    return max(0.0, 1.0 - ($avgDiff / 255.0));
}

function jps_grayscale_grid(\GdImage $image, int $gridSize): array
{
    [$w, $h] = jps_image_size($image);
    $small = imagecreatetruecolor($gridSize, $gridSize);
    imagecopyresampled($small, $image, 0, 0, 0, 0, $gridSize, $gridSize, $w, $h);

    $values = [];
    for ($y = 0; $y < $gridSize; $y++) {
        for ($x = 0; $x < $gridSize; $x++) {
            $rgb = imagecolorat($small, $x, $y);
            $r = ($rgb >> 16) & 0xFF;
            $g = ($rgb >> 8) & 0xFF;
            $b = $rgb & 0xFF;
            $values[] = 0.299 * $r + 0.587 * $g + 0.114 * $b;
        }
    }
    imagedestroy($small);
    return $values;
}

/**
 * @return array{image: \GdImage|null, used_cache: bool, from_gemini: bool, similarity: float|null, is_safe: bool, error: string|null}
 */
function jps_gemini_edit_image(\GdImage $image, string $mode, string $backgroundStyle, string $cacheDir): array
{
    $result = ['image' => null, 'used_cache' => false, 'from_gemini' => false, 'similarity' => null, 'is_safe' => false, 'error' => null];

    if (!jps_gemini_configured()) {
        $result['error'] = 'Gemini is not configured (missing GEMINI_API_KEY).';
        return $result;
    }

    $uploadImage = jps_resize_max_dimension(jps_clone_image($image), GEMINI_MAX_UPLOAD_DIMENSION);
    $instruction = jps_build_gemini_instruction($mode, $backgroundStyle);

    ob_start();
    imagepng($uploadImage);
    $pngBytes = ob_get_clean();

    if (!is_dir($cacheDir)) {
        @mkdir($cacheDir, 0775, true);
    }
    $cacheKey = hash('sha256', $pngBytes . $instruction . GEMINI_IMAGE_MODEL);
    $cachePath = $cacheDir . '/' . $cacheKey . '.png';

    if (is_readable($cachePath)) {
        $edited = imagecreatefrompng($cachePath);
        if ($edited !== false) {
            $similarity = jps_image_similarity($uploadImage, $edited);
            $result['image'] = $edited;
            $result['used_cache'] = true;
            $result['from_gemini'] = true;
            $result['similarity'] = $similarity;
            $result['is_safe'] = $similarity >= GEMINI_MIN_SIMILARITY;
            return $result;
        }
    }

    try {
        $rawBytes = jps_gemini_call_with_retries($pngBytes, $instruction);
    } catch (\Throwable $e) {
        $result['error'] = $e->getMessage();
        return $result;
    }

    $edited = @imagecreatefromstring($rawBytes);
    if ($edited === false) {
        $result['error'] = 'Gemini response did not contain a decodable image.';
        return $result;
    }

    $similarity = jps_image_similarity($uploadImage, $edited);
    @imagepng($edited, $cachePath);

    $result['image'] = $edited;
    $result['from_gemini'] = true;
    $result['similarity'] = $similarity;
    $result['is_safe'] = $similarity >= GEMINI_MIN_SIMILARITY;
    return $result;
}

function jps_clone_image(\GdImage $image): \GdImage
{
    [$w, $h] = jps_image_size($image);
    $copy = imagecreatetruecolor($w, $h);
    imagecopy($copy, $image, 0, 0, 0, 0, $w, $h);
    return $copy;
}

/** @throws Exception */
function jps_gemini_call_with_retries(string $pngBytes, string $instruction): string
{
    $lastError = null;
    for ($attempt = 1; $attempt <= GEMINI_MAX_RETRIES; $attempt++) {
        try {
            return jps_gemini_call_once($pngBytes, $instruction);
        } catch (\Exception $e) {
            $lastError = $e;
            if ($attempt >= GEMINI_MAX_RETRIES) {
                break;
            }
            $message = strtolower($e->getMessage());
            $backoff = GEMINI_RETRY_BACKOFF_SECONDS * (2 ** ($attempt - 1));
            if (str_contains($message, 'rate') || str_contains($message, '429') || str_contains($message, 'resource_exhausted')) {
                usleep((int) ($backoff * 2 * 1_000_000));
            } else {
                usleep((int) ($backoff * 1_000_000));
            }
        }
    }
    throw $lastError ?? new Exception('Unknown Gemini error.');
}

/** @throws Exception */
function jps_gemini_call_once(string $pngBytes, string $instruction): string
{
    $url = sprintf(
        'https://generativelanguage.googleapis.com/v1beta/models/%s:generateContent?key=%s',
        rawurlencode(GEMINI_IMAGE_MODEL),
        rawurlencode(GEMINI_API_KEY)
    );

    $payload = [
        'contents' => [[
            'parts' => [
                ['text' => $instruction],
                ['inline_data' => ['mime_type' => 'image/png', 'data' => base64_encode($pngBytes)]],
            ],
        ]],
    ];

    $ch = curl_init($url);
    curl_setopt_array($ch, [
        CURLOPT_POST => true,
        CURLOPT_POSTFIELDS => json_encode($payload),
        CURLOPT_HTTPHEADER => ['Content-Type: application/json'],
        CURLOPT_RETURNTRANSFER => true,
        CURLOPT_TIMEOUT => GEMINI_TIMEOUT_SECONDS,
        CURLOPT_CONNECTTIMEOUT => 15,
    ]);
    $response = curl_exec($ch);
    $httpCode = curl_getinfo($ch, CURLINFO_HTTP_CODE);
    $curlErrno = curl_errno($ch);
    $curlError = curl_error($ch);
    curl_close($ch);

    if ($curlErrno === CURLE_OPERATION_TIMEDOUT) {
        throw new Exception('Gemini request timed out.');
    }
    if ($response === false) {
        throw new Exception("Gemini request failed: $curlError");
    }
    if ($httpCode === 429) {
        throw new Exception('Gemini rate limit hit (429).');
    }
    if ($httpCode >= 500) {
        throw new Exception("Gemini server error ($httpCode).");
    }
    if ($httpCode !== 200) {
        $snippet = substr($response, 0, 300);
        throw new Exception("Gemini request failed with HTTP $httpCode: $snippet");
    }

    $data = json_decode($response, true);
    if (!is_array($data)) {
        throw new Exception('Gemini returned an unparsable response.');
    }

    $parts = $data['candidates'][0]['content']['parts'] ?? [];
    foreach ($parts as $part) {
        $inline = $part['inlineData'] ?? $part['inline_data'] ?? null;
        if (is_array($inline) && !empty($inline['data'])) {
            $decoded = base64_decode($inline['data'], true);
            if ($decoded !== false) {
                return $decoded;
            }
        }
    }

    throw new Exception('Gemini response did not contain an edited image.');
}

function jps_estimate_api_usage_warning(int $numImages, bool $useGemini): ?string
{
    if (!$useGemini || $numImages === 0) {
        return null;
    }
    return "This batch will send up to $numImages image edit request(s) to the Gemini API " .
        '(fewer if some are served from cache). Large batches may take a while and are subject ' .
        "to your Google AI Studio / Gemini API usage quota and billing.";
}
